src/batch/src/executor2/hash_agg.rs

```diff
@@ -192,66 +207,50 @@ impl<K: HashKey + Send + Sync> Executor for HashAggExecutor<K> {
         assert!(self.result.is_none());
         self.result = Some(mem::take(&mut self.groups).into_iter());
 
-        Ok(())
-    }
-
-    async fn next(&mut self) -> Result<Option<DataChunk>> {
+        // generate output data chunks
         if let Some(res) = self.result.as_mut() {
-            let cardinality = DEFAULT_CHUNK_BUFFER_SIZE;
-            let mut group_builders = self
-                .group_key_types
-                .iter()
-                .map(|datatype| datatype.create_array_builder(cardinality))
-                .collect::<Result<Vec<_>>>()?;
-
-            let mut agg_builders = self
-                .agg_factories
-                .iter()
-                .map(|agg_factory| {
-                    agg_factory
-                        .get_return_type()
-                        .create_array_builder(cardinality)
-                })
-                .collect::<Result<Vec<_>>>()?;
-
-            let mut has_next = false;
-            for (key, states) in res.take(cardinality) {
-                has_next = true;
-                key.deserialize_to_builders(&mut group_builders[..])?;
-                states
+            loop {
+                let cardinality = DEFAULT_CHUNK_BUFFER_SIZE;
+                let mut group_builders = self
+                    .group_key_types
+                    .iter()
+                    .map(|datatype| datatype.create_array_builder(cardinality))
+                    .collect::<Result<Vec<_>>>()?;
+
+                let mut agg_builders = self
+                    .agg_factories
+                    .iter()
+                    .map(|agg_factory| {
+                        agg_factory
+                            .get_return_type()
+                            .create_array_builder(cardinality)
+                    })
+                    .collect::<Result<Vec<_>>>()?;
```

Comment:
Is it possible to reuse these array builders (`group_builders` and `agg_builders`) across the loops?

Reply:
I also notice that. But if we move the builders outside the loop, it will incur some ownership problems.

Reply:
And each time we build an output chunk, we need to deserialize the hash key into builders, seems that we cannot reuse them.